src/Palette.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2023 Inclusive Design Research Centre, OCAD University
+ * All rights reserved.
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ *
+ * You may obtain a copy of the License at
+ * https://github.com/inclusive-design/adaptive-palette/blob/main/LICENSE
+ */
+
+import { render } from "preact";
+import { html } from "htm/preact";
+import { PaletteCell } from "./PaletteCell";
+
+/**
+ * Given a palette defined in a json structure, compute the number of rows
+ * and columns in that palette.
+ *
+ * @param {Object} paletteDefinition - An object that lists the positions,
+ *                 heights and widths of the cells in the palette.
+ * @return {Object} - The row and column counts: `{ numRows: ..., numColumns: ...}`.
+ */
+function countRowsColumns (paletteDefinition) {
+  let rowCount = 0;
+  let colCount = 0;
+  let rightColumn = 0;
+  let bottomRow = 0;
+  const cellIds = Object.keys(paletteDefinition.cells);
+  cellIds.forEach((id) => {
+    const aCell = paletteDefinition.cells[id];
+    rightColumn = aCell.columnStart + aCell.columnSpan;
+    if (rightColumn > colCount) {
+      colCount = rightColumn;
+    }
+    bottomRow = aCell.rowStart + aCell.rowSpan;
+    if (bottomRow > rowCount) {
+      rowCount = bottomRow;
+    }
+  });
+  return { numRows: rowCount, numColumns: colCount };
+}
+
+export function Palette (props) {
+
+  const paletteDefinition = props.json;
+  const rowsCols = countRowsColumns(paletteDefinition);
+  const cellIds = Object.keys(paletteDefinition.cells);
+
+  // Loop to create an array of renderings for each cell
+  const theCells = [];
+  cellIds.forEach((id) => {
+    const aCell = paletteDefinition.cells[id];
+    const paletteCell = html`
+      <${PaletteCell} id="${id}" labelText="${aCell.label}"
```

Comment:
A palette cell can be one of types: ActionBmwIcon, ContentBmwEncoding etc.

I was thinking the `type` field defined for each cell specifies what type it is, and a component named in this type name can be instantiated to render this cell. If we use this approach, `PaletteCell` component should be renamed to `ActionBmwIcon` (let me know if you come up with a better name), and the `aCell.type` value should replace `${PaletteCell}` in this line.

I'm also thinking each cell type has an `options` block in the JSON for holding the cell specific information. This `options` can be directly when this cell gets instantiated so the instantiator at this line doen't need to worry about what property values are expected. The property structure can be verified inside the cell component. If we do so, the JSON structure for `ActionBmwIcon` will look like:

```
{
    "type": "ActionBmwKey",
    "rowStart": 4,
    "rowSpan": 1,
    "columnStart": 1,
    "columnSpan": 1,
    "options": {
        "label": "ADJ.+ER",
        "BCI-AV-ID": 24879
    }
}
```

Let me know what you think. Thanks.

Reply:
> A palette cell can be one of types: ActionBmwIcon, ContentBmwEncoding etc.

And also "Branch" and "BranchBack" for navigating among different layers.  And others we haven't thought of yet.

> I was thinking the type field defined for each cell specifies what type it is,
> and a component named in this type name can be instantiated to render this
> cell. If we use this approach, PaletteCell component should be renamed to
> ActionBmwIcon (let me know if you come up with a better name), and the
> aCell.type value should replace ${PaletteCell} in this line.

I have been thinking about this, and I keep coming back to using classes since that is a technique for putting common aspects of all palette cells in a base component and deriving more specialized cases therefrom.  I'm not sure how to do that with Functional Preact components, but perhaps there is a way.  Maybe it's as simple as calling "base" Functions from "derived" Functions.  And, in that case, maybe `Function PaletteCell()` is the most basic in the chain.

Another thought I had about the `type` field was to put it into a `data-cellType` attribute in the html itself since all of the action cells are `button` elements.  Without the `data-cellType` attribute, it is impossible to tell what kind of `button` it is (unless the `class` attribute has sufficient information?).  My intuition is that "Undo" or "Delete" are different types of actions than entering a BMW code. But, I don't know yet if there is any use to noting the cell type in the markup.  I'll file away the idea for now.

> I'm also thinking each cell type has an options block in the JSON for holding
> the cell specific information. This options can be directly when this cell
> gets instantiated so  ...

There is a missing verb here:  "... can be directly " ? Copied?  I think you mean something like this in terms of using the `options` block: `<$PaletteCell options=${aCell.options} id=${id} ...`.

Then the `PaletteCell` component, or sub-component, uses only the options that are appropriate for that type.

That's a great idea.  In that case, I am inclined to put the row/column info in the options block as well.